Src/Support/GoogleApis.Core/Apis/Logging/BaseLogger.cs

```diff
@@ -0,0 +1,182 @@
+﻿/*
+Copyright 2011 Google Inc
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+using Google.Apis.Util;
+using System;
+using System.Globalization;
+
+namespace Google.Apis.Logging
+{
+    /// <summary>
+    /// An abstract base logger, upon which real loggers may be built.
+    /// </summary>
+    public abstract class BaseLogger : ILogger
+    {
+        // Matches gRPC datetime log format
+        private const string DateTimeFormatString = "MMdd HH:mm:ss.ffffff";
+
+        /// <summary>
+        /// Construct a <see cref="BaseLogger"/>. 
+        /// </summary>
+        /// <param name="minimumLogLevel">Logging will be enabled at this level and all higher levels.</param>
+        /// <param name="clock">The <see cref="IClock"/> to use to timestamp log entries.</param>
+        /// <param name="forType">The type from which entries are being logged. May be <c>null</c>.</param>
+        protected BaseLogger(LogLevel minimumLogLevel, IClock clock, Type forType)
+        {
+            MinimumLogLevel = minimumLogLevel;
+            IsDebugEnabled = minimumLogLevel <= LogLevel.Debug;
+            IsInfoEnabled = minimumLogLevel <= LogLevel.Info;
+            IsWarningEnabled = minimumLogLevel <= LogLevel.Warning;
+            IsErrorEnabled = minimumLogLevel <= LogLevel.Error;
+            Clock = clock ?? SystemClock.Default;
+            LoggerForType = forType;
+            if (forType != null)
+            {
+                var namespaceStr = forType.Namespace ?? "";
+                if (namespaceStr.Length > 0)
+                {
+                    namespaceStr += ".";
+                }
+                _loggerForTypeString = namespaceStr + forType.Name + " ";
+            }
+            else
+            {
+                _loggerForTypeString = "";
+            }
+        }
+
+        private readonly string _loggerForTypeString;
+
+        /// <summary>
+        /// The <see cref="IClock"/> being used to timestamp log entries. 
+        /// </summary>
+        public IClock Clock { get; }
+
+        /// <summary>
+        /// The type from which entries are being logged. May be <c>null</c>.
+        /// </summary>
+        public Type LoggerForType { get; }
```

Comment:
What's the benefit of this? I've seen this in gRPC, but it seems to add more complexity than value.